File: projects/DotNet/CSharp/Tests/HelloTests.cs
using NSubstitute;
using NUnit.Framework;

namespace HelloWorld.Tests
{
    [TestFixture]
    public class HelloTests
    {
        [Test]
        public void FirstPassingTest()
        {
            Assert.True(true);
        }

        [Test]
        public void FirstFailingTest()
        {
            Assert.False(true, "I should fail!!");
        }

        [Test]
        public void StubTestWithNSubstitute()
        {
            //Arrange
            var service = Substitute.For<IWeatherService>();
            service.GetWeather("Rome").Returns("Sunny");
            service.GetWeather("Cortina").Returns("Snowy");

            //Act
            var actual = service.GetWeather("Rome");

            //Assert
            Assert.AreEqual("Sunny", actual);
        }

        [Test]
        public void MockTestWithNSubstitute()
        {
            //Arrange
            var service = Substitute.For<IWeatherService>();
            service.GetWeather("Rome").Returns("Sunny");
            service.GetWeather("Cortina").Returns("Snowy");

            //Act
            var actual = service.GetWeather("Rome");

            //Assert
            Assert.AreEqual("Sunny", actual);

            service.Received().GetWeather("Rome");
            service.DidNotReceive().GetWeather("Cortina");
        }
    }
}

File: projects/DotNet/VB.NET/HelloWorld/HelloTests.vb
Imports NUnit.Framework
Imports NSubstitute

<TestFixture()> _
Public Class HelloTests

    <Test()> _
    Public Sub FirstPassingTest()
        Assert.True(True)
    End Sub

    '<Test()> _
    'Public Sub FirstFailingTest()
    '    Assert.False(True, "I should fail!!")
    'End Sub

    <Test()> _
    Public Sub StubTestWithNSubstitute()

        'Arrange
        Dim service = Substitute.For(Of IWeatherService)()
        service.GetWeather("Rome").Returns("Sunny")
        service.GetWeather("Cortina").Returns("Snowy")

        'Act
        Dim actual As String = service.GetWeather("Rome")

        'Assert
        Assert.AreEqual("Sunny", actual)
    End Sub

    <Test()> _
    Public Sub MockTestWithNSubstitute()

        'Arrange
        Dim service = Substitute.For(Of IWeatherService)()
        service.GetWeather("Rome").Returns("Sunny")
        service.GetWeather("Cortina").Returns("Snowy")

        'Act
        Dim actual As String = service.GetWeather("Rome")

        'Assert
        Assert.AreEqual("Sunny", actual)

        service.Received().GetWeather("Rome")
        service.DidNotReceive().GetWeather("Cortina")
    End Sub


End Class


